File: Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/data_preprocessing.py
# Data Preprocessing TEMPLATE

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


# Importing the dataset
dataset = pd.read_csv('Data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 3].values

# taking care of missing values 
from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
# not we fit the obect to matrix x 
imputer = imputer.fit(X[:, 1:3])
# replace the missing data with the mean 
X[:, 1:3] = imputer.transform(X[:, 1:3])


# Encoding Catagorical Data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
# calling labelencoder class with no arguments passed as 
labelencoder_X = LabelEncoder()
# now we are using this object on the desired variable 
# and we are assigning the values to each 
X[:, 0] = labelencoder_X.fit_transform(X[:, 0])
# but the issue is still the same. 
# since 1 is > 0 and 2>1 the program will think that these countries are of higher vlaue
# so we have to prevent the machine from thinking this via dummy variables 
# We have to specify the index that is the column of the catagorical which is `0`
onehotencoder = OneHotEncoder(categorical_features = [0])
X = onehotencoder.fit_transform(X).toarray()
# Now we take care of the purchased column 
 # We will just need to use label encoder because it is the dependant variable 
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)

# Splitting the dataset into the training and test set
# we are going to import the cross validation library
from sklearn.cross_validation import train_test_split 
# now we build the training and test sets 14HBLmX1xEQmzNCy843qE3YkWUZAfWj2yX
# under train_test_split we first put the independent vriables 'x' in this case
# then we do the dependant variable 'y' in this case
# then we set the test size so .5 would be half goes to training set and half to test
# test size is usualy 20-25% 
# so here we will have two observations in test set and 8 in the training set 
# random_state is included for the course so we have the same logical results 
X_train, X_test, y_train, y_test  = train_test_split(X, y, test_size = 0.2, random_state = 0)
# so we build a machine learning model by building relatioship from the independent model to the dependent model 
# then we test if the model can be applied or corelated on the test set 

# FEATURE SCALLING 
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
# fit transformation to data then transform it 
# We fit the training data so we can build the model off of it 
X_train = sc_X.fit_transform(X_train)
# Perform standardization by centering the scaling 
# x = x - mu (which is the mean) / standard deviation
# the below is the same since we already fit the model we can just tranform 
# We do not fit the test data as we want to predict 
X_test = sc_X.transform(X_test)

File: Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/data_preprocessing.R
# Data Preprocessing Template

# Importing the dataset
dataset = read.csv('Data.csv')

# Taking care of missing data
# is.na checks to see if all the values in the specified column are missing
# na.rm to compute the missing value 
# the third ',' is for is the value is not missing
dataset$Age = ifelse(is.na(dataset$Age), ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)), dataset$Age)

# repeat for salary 
dataset$Salary = ifelse(is.na(dataset$Salary),ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),dataset$Salary)

# Encoding the country column
# 'c' is a vector that will contain the countries 
# then we choose the label we assign to the countries 
dataset$Country = factor(dataset$Country,
                         levels = c('France', 'Spain', 'Germany'),
                         labels = c(1, 2, 3))

# Now we repeat    
dataset$Purchased = factor(dataset$Purchased,
                         levels = c('No', 'Yes'),
                         labels = c(0, 1))

# splitting dataset into training set 
# we has to install using         install.packages('caTools') 
# then we had to go to packages tag and select or we could     library(caTools)
# instead of using random state like in python we are going to do it manually through seed with a random  number 
set.seed(42)
# Now we split it 
# We use the $Purchased as this is the dependent variable 
# split ratio is the percentag ewe want to put in TRAINING set. We did 20% in python so we will do the same here)
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# Now we can run `split` and see true means it goes to training set and false to the test set
# Now we can create the sets differently. Each set is a subset of the data set
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)


#Feature Scaling 
# using the below causes issues because we transfomed the words/factors into numbers 
# training_set = scale(training_set)
# test_set = scale(test_set)
# so we are going to exclude these
training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3])
